src/panels/lovelace/editor/card-editor/hui-card-editor.ts

```diff
@@ -0,0 +1,264 @@
+  public get config(): LovelaceCardConfig | undefined {
+    try {
+      return yaml.safeLoad(this.yaml);
+    } catch (err) {
+      this._error = err;
+      return { type: "error", error: err.message };
+    }
+  }
+  public set config(config: LovelaceCardConfig | undefined) {
+    this.yaml = yaml.safeDump(config);
+  }
+
+  public get error(): boolean {
+    return this._error !== undefined;
+  }
+
+  private get _yamlEditor(): HuiYamlEditor {
+    return this.shadowRoot!.querySelector("hui-yaml-editor")!;
+  }
+
+  protected render(): TemplateResult {
+    return html`
+      <div style="width: 100%;">
+        <div class="gui-editor" ?hidden="${this._GUImode === false}">
+          ${this._loading
+            ? html`
+                <paper-spinner
+                  ?active="${this._loading}"
+                  alt="Loading"
+                  class="center margin-bot"
+                ></paper-spinner>
+              `
+            : this._warning || this._error
+            ? ""
+            : this._configElement}
+        </div>
+        <div class="yaml-editor" ?hidden="${this._GUImode === true}">
+          <hui-yaml-editor
+            .hass="${this.hass}"
+            .value="${this.yaml}"
+            @yaml-changed="${this._handleYAMLChanged}"
+          ></hui-yaml-editor>
+          <div class="error">
+            ${this._error}
+          </div>
+          <div class="warning">
+            ${this._warning}
+          </div>
+        </div>
+        <div class="buttons">
+          <mwc-button
+            @click="${() => (this._GUImode = !this._GUImode)}"
+            ?disabled="${this._warning || this._error}"
+            ?unelevated="${this._GUImode === false}"
+          >
+            <ha-icon .icon="${"mdi:code-braces"}"></ha-icon>
+          </mwc-button>
+        </div>
+      </div>
+    `;
+  }
+
+  static get styles(): CSSResult {
+    return css`
+      :host {
+        margin: 0 10px;
+        display: flex;
+      }
+      .hidden {
+        display: none;
+      }
+      .gui-editor,
+      .yaml-editor {
+        padding: 8px 0px;
+      }
+      .error {
+        color: #ef5350;
+      }
+      .warning {
+        color: #ffa726;
+      }
+      .buttons {
+        text-align: right;
+
+        padding: 8px 0px;
+        border-top: 1px solid #e8e8e8;
+      }
+    `;
+  }
+
+  protected async firstUpdated() {
+    this.config = this.value;
+  }
+
+  protected updated(changedProperties) {
+    super.updated(changedProperties);
+
+    if (changedProperties.has("_GUImode")) {
+      if (this._GUImode === false) {
+        // Refresh code editor when switching to yaml mode
+        this._yamlEditor.codemirror.refresh();
+        this._yamlEditor.codemirror.focus();
+      }
+      fireEvent(this as HTMLElement, "iron-resize");
+    }
+  }
+
+  private _handleUIConfigChanged(ev) {
+    ev.stopPropagation();
+    const config = ev.detail.config;
+    if (JSON.stringify(config) !== JSON.stringify(this.config)) {
```

Comment:
why would we do this expensive computation ? We should just apply it.

Reply:
My thinking was this could be used to check for changes that needed saving. But that's better handled somewhere else... and also a thing that I'd forgotten about entirely...